### PR TITLE
Update pointcloud.py

### PR DESCRIPTION
**Pull request by mitchelldaneker**
allow PointCloud geometry to support defining boundary points and normals.

### CODE REVIEW

File: deepxde/geometry/pointcloud.py [hunk 1]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
 
     Args:
         points: A NumPy array of shape (`N`, `d`). A list of `d`-dim points.
+        If boundary is known, includes only points inside the boundary.
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:13]: indent

File: deepxde/geometry/pointcloud.py [hunk 2]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
...
+        boundary_points: A NumPy array of shape (`N`, `d`). A list of `d`-dim
+        points on the boundary. If boundary is not known, leave as None type.
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:15]: indent

File: deepxde/geometry/pointcloud.py [hunk 3]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
...
+        normals: A NumPy array of shape (`N`, `d`). A list of `d`-dim
+        points of the normals corresponding to boundary_points.
     """
 
-    def __init__(self, points):
+    def __init__(self, points, boundary_points=None, normals=None):
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:20]: normals --> boundary_normals

**Comment by lululxvi** [comment]
Let me mark resolved in the future.

File: deepxde/geometry/pointcloud.py [hunk 4]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
 
     Args:
         points: A NumPy array of shape (`N`, `d`). A list of `d`-dim points.
+            If boundary is known, includes only points inside the boundary.
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:13]: If `boundary_points` is provided, includes only points inside the geometry.

File: deepxde/geometry/pointcloud.py [hunk 5]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
 
     Args:
         points: A NumPy array of shape (`N`, `d`). A list of `d`-dim points.
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:12]: A NumPy array of shape (`N`, `d`). --> A 2-D NumPy array.

File: deepxde/geometry/pointcloud.py [hunk 6]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
...
+            If boundary is known, includes only points inside the boundary.
+        boundary_points: A NumPy array of shape (`N`, `d`). A list of `d`-dim
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:14]: A NumPy array of shape (`N`, `d`). --> A 2-D NumPy array.

File: deepxde/geometry/pointcloud.py [hunk 7]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
...
+            points on the boundary. If boundary is not known, leave as None type.
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:15]: Delete "If boundary is not known, leave as None type."

File: deepxde/geometry/pointcloud.py [hunk 8]
```diff
@@ -10,24 +10,73 @@ class PointCloud(Geometry):
...
+        normals: A NumPy array of shape (`N`, `d`). A list of `d`-dim
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:16]: Also update here.

File: deepxde/geometry/pointcloud.py (hunk 5 above)
Comment (lululxvi) [reply deepxde/geometry/pointcloud.py]: I mean replacing "A NumPy array of shape (`N`, `d`)." with "A 2-D NumPy array."

File: deepxde/geometry/pointcloud.py (hunk 6 above)
Comment (lululxvi) [reply deepxde/geometry/pointcloud.py]: I mean replacing "A NumPy array of shape (N, d)." with "A 2-D NumPy array."

File: deepxde/geometry/pointcloud.py [hunk 9]
```diff
@@ -9,25 +9,74 @@ class PointCloud(Geometry):
...
     """
 
-    def __init__(self, points):
+    def __init__(self, points, boundary_points=None, boundary_normals=None):
         self.points = np.asarray(points, dtype=config.real(np))
         self.num_points = len(points)
+        if boundary_points is None:
+            if boundary_normals is not None:
+                raise ValueError(
+                    "boundary_points must be provided to use boundary_normals"
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:26]: No. boundary_normals can be None if not using normals.

File: deepxde/geometry/pointcloud.py [hunk 10]
```diff
@@ -9,25 +9,74 @@ class PointCloud(Geometry):
...
+                )
+            self.boundary_points = boundary_points
+            self.boundary_normals = boundary_normals
+            self.all_points = np.asarray(points, dtype=config.real(np))
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:30]: self.all_points = self.points

File: deepxde/geometry/pointcloud.py [hunk 11]
```diff
@@ -9,25 +9,74 @@ class PointCloud(Geometry):
...
+        else:
+            if boundary_normals is not None:
+                if boundary_normals.shape != boundary_points.shape:
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:33]: shape is a numpy array. 
```
if len(boundary_normals) != len(boundary_points):
```

File: deepxde/geometry/pointcloud.py [hunk 12]
```diff
@@ -9,25 +9,74 @@ class PointCloud(Geometry):
...
+                    raise ValueError(
+                        "the shape of boundary_normals should be the same as boundary_points"
+                    )
+            self.boundary_normals = boundary_normals
+            self.boundary_points = np.asarray(
+                boundary_points, dtype=config.real(np)
+            )
+            self.num_boundary_points = len(boundary_points)
+            self.all_points = np.vstack(
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:42]: self.all_points = np.vstack((self.points, self.boundary_points))

File: deepxde/geometry/pointcloud.py [hunk 13]
```diff
@@ -9,25 +9,74 @@ class PointCloud(Geometry):
...
         )
-
         self.sampler = BatchSampler(self.num_points, shuffle=True)
 
     def inside(self, x):
-        raise NotImplementedError("dde.geometry.PointSet doesn't support inside.")
+        x_broad = x[:, None, :]
+        broad = self.points[None, :, :]
+        return (
+            np.isclose((x_broad - broad), 0, atol=1e-6).all(axis=2).any(axis=1)
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:65]: np.isclose((x[:, None, :] - self.points[None, :, :]), 0, atol=1e-6).all(axis=2).any(axis=1)

File: deepxde/geometry/pointcloud.py [hunk 14]
```diff
@@ -9,25 +9,74 @@ class PointCloud(Geometry):
...
+        )
 
     def on_boundary(self, x):
-        raise NotImplementedError("dde.geometry.PointSet doesn't support on_boundary.")
+        if self.boundary_points is None:
+            raise ValueError(
+                "boundary_points must be defined to sample the boundary"
+            )
+        x_broad = x[:, None, :]
+        bound_broad = self.boundary_points[None, :, :]
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:74]: No need to define x_broad and bound_broad.

File: deepxde/geometry/pointcloud.py [hunk 15]
```diff
@@ -9,25 +9,58 @@ class PointCloud(Geometry):
     """A geometry represented by a point cloud, i.e., a set of points in space.
 
     Args:
-        points: A NumPy array of shape (`N`, `d`). A list of `d`-dim points.
+        points: A 2-D NumPy array. If `boundary_points` is not provided, `points` can 
+            include points both inside the geometry or on the boundary; if `boundary_points`
+            is provided, `points` includes only points inside the geometry.
+        boundary_points: A 2-D NumPy array.
+        normals: A 2-D NumPy array.
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:16]: normals --> boundary_normals

File: deepxde/geometry/pointcloud.py [hunk 16]
```diff
@@ -9,25 +9,58 @@ class PointCloud(Geometry):
...
         self.points = np.asarray(points, dtype=config.real(np))
         self.num_points = len(points)
+        if boundary_points is None:
+            if boundary_normals is not None:
+                raise ValueError(
+                    "boundary_points must be provided to use boundary_normals"
+                )
+            self.boundary_points = boundary_points
+            self.boundary_normals = boundary_normals
+            self.all_points = self.points
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:29]: `self.all_points` --> `all_points`. No need to save in self. It is a waste of memory.

File: deepxde/geometry/pointcloud.py [hunk 17]
```diff
@@ -9,25 +9,58 @@ class PointCloud(Geometry):
...
+    def __init__(self, points, boundary_points=None, boundary_normals=None):
         self.points = np.asarray(points, dtype=config.real(np))
         self.num_points = len(points)
+        if boundary_points is None:
+            if boundary_normals is not None:
+                raise ValueError(
+                    "boundary_points must be provided to use boundary_normals"
+                )
+            self.boundary_points = boundary_points
+            self.boundary_normals = boundary_normals
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:28]: self.boundary_points = None
self.boundary_normals = None

File: deepxde/geometry/pointcloud.py [hunk 18]
```diff
@@ -9,25 +9,58 @@ class PointCloud(Geometry):
...
+            if boundary_normals is not None:
+                if len(boundary_normals) != len(boundary_points):
+                    raise ValueError(
+                        "the shape of boundary_normals should be the same as boundary_points"
+                    )
+            self.boundary_normals = boundary_normals
+            self.boundary_points = np.asarray(
+                boundary_points, dtype=config.real(np)
+            )
+            self.num_boundary_points = len(boundary_points)
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:40]:
```
self.boundary_points = np.asarray(boundary_points, dtype=config.real(np))
self.num_boundary_points = len(boundary_points)
self.boundary_normals = boundary_normals
```

File: deepxde/geometry/pointcloud.py [hunk 19]
```diff
@@ -9,25 +9,58 @@ class PointCloud(Geometry):
...
 
     def inside(self, x):
-        raise NotImplementedError("dde.geometry.PointSet doesn't support inside.")
+        return np.isclose((x[:, None, :] - self.points[None, :, :]), 0, atol=1e-6).all(axis=2).any(axis=1)
 
     def on_boundary(self, x):
-        raise NotImplementedError("dde.geometry.PointSet doesn't support on_boundary.")
+        if self.boundary_points is None:
+            raise ValueError(
+                "boundary_points must be defined to sample the boundary"
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:61]: boundary_points must be defined to test if a point is on the boundary

Comment (mitchelldaneker) [reply deepxde/geometry/pointcloud.py:61]: Should this be: 

`boundary_points` must be defined to test `on_boundary`

Comment (lululxvi) [reply deepxde/geometry/pointcloud.py:61]: Simply use this:
"boundary_points must be defined to test on_boundary"

File: deepxde/geometry/pointcloud.py (hunk 18 above)
Comment (mitchelldaneker) [reply deepxde/geometry/pointcloud.py:37]: This appears to me in the format you have already suggested

File: deepxde/geometry/pointcloud.py [hunk 20]
```diff
@@ -9,25 +9,56 @@ class PointCloud(Geometry):
...
+        if boundary_points is None:
+            if boundary_normals is not None:
+                raise ValueError(
+                    "boundary_points must be provided to use boundary_normals"
+                )
+            self.boundary_points = None
+            self.boundary_normals = None
+            all_points = self.points
+        else:
+            if boundary_normals is not None:
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:31]: if boundary_normals is not None and len(boundary_normals) != len(boundary_points):
    raise ...

File: deepxde/geometry/pointcloud.py [hunk 21]
```diff
@@ -9,25 +9,56 @@ class PointCloud(Geometry):
...
+                if len(boundary_normals) != len(boundary_points):
+                    raise ValueError(
+                        "the shape of boundary_normals should be the same as boundary_points"
+                    )
+            self.boundary_points = np.asarray(boundary_points, dtype=config.real(np))
+            self.num_boundary_points = len(boundary_points)
+            self.boundary_normals = boundary_normals
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:38]: Need to use `np.asarray` for normals as well if it is not None.

Comment (mitchelldaneker) [reply deepxde/geometry/pointcloud.py]: I have implemented this

File: deepxde/geometry/pointcloud.py [hunk 22]
```diff
@@ -9,25 +9,58 @@ class PointCloud(Geometry):
...
+            if boundary_normals is not None:
+                if len(boundary_normals) != len(boundary_points):
+                    raise ValueError(
+                        "the shape of boundary_normals should be the same as boundary_points"
+                    )
+                self.boundary_normals = np.asarray(boundary_normals, dtype=config.real(np))
+            else:
+                self.boundary_normals = boundary_normals
+            self.boundary_points = np.asarray(boundary_points, dtype=config.real(np))
+            self.num_boundary_points = len(boundary_points)
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:40]:
```
self.boundary_points = np.asarray(boundary_points, dtype=config.real(np))
self.num_boundary_points = len(boundary_points)
if boundary_normals is not None:
    ...
else:
    self.boundary_normals = None
```

File: deepxde/geometry/pointcloud.py [hunk 23]
```diff
@@ -9,25 +9,58 @@ class PointCloud(Geometry):
...
+            is provided, `points` includes only points inside the geometry.
+        boundary_points: A 2-D NumPy array.
+        boundary_normals: A 2-D NumPy array.
     """
 
-    def __init__(self, points):
+    def __init__(self, points, boundary_points=None, boundary_normals=None):
         self.points = np.asarray(points, dtype=config.real(np))
         self.num_points = len(points)
+        if boundary_points is None:
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:22]:
```
self.num_points = len(points)
self.boundary_points = None
self.boundary_normals = None
```
Then delete all the lines below of `self.boundary_points = None` and `self.boundary_normals = None`

Comment (mitchelldaneker) [reply deepxde/geometry/pointcloud.py:22]: I am confused by what you mean. Do you mean to do something like the following?

```
self.points = np.asarray(points, dtype=config.real(np))
self.num_points = len(points) 
self.boundary_points = None 
self.boundary_normals = None
all_points = self.points
if boundary_points is not None:
        self.boundary_points = np.asarray(boundary_points, dtype=config.real(np))
        self.num_boundary_points = len(boundary_points)
        all_points = np.vstack((self.points, self.boundary_points))
        self.boundary_sampler = BatchSampler(
            self.num_boundary_points, shuffle=True
        )
        if boundary_normals is not None:
                        if len(boundary_normals) != len(boundary_points):
                            raise ValueError(
                                "the shape of boundary_normals should be the same as boundary_points"
                            )
                        self.boundary_normals = np.asarray(boundary_normals, dtype=config.real(np))
super().__init__(
            len(points[0]),
            (
                np.amin(all_points, axis=0),
                np.amax(all_points, axis=0),
            ),
            np.inf,
        )
```

Comment (lululxvi) [reply deepxde/geometry/pointcloud.py:22]: yes

File: deepxde/geometry/pointcloud.py [hunk 24]
```diff
@@ -9,25 +9,51 @@ class PointCloud(Geometry):
...
-    def __init__(self, points):
+    def __init__(self, points, boundary_points=None, boundary_normals=None):
         self.points = np.asarray(points, dtype=config.real(np))
-        self.num_points = len(points)
+        self.num_points = len(points) 
+        self.boundary_points = None 
+        self.boundary_normals = None
+        all_points = self.points
+        if boundary_points is not None:
+                self.boundary_points = np.asarray(boundary_points, dtype=config.real(np))
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:26]: The indent should be 4 spaces

Comment (mitchelldaneker) [reply deepxde/geometry/pointcloud.py]: Should be fixed now.

File: deepxde/geometry/pointcloud.py [hunk 25]
```diff
@@ -9,25 +9,51 @@ class PointCloud(Geometry):
...
+            is provided, `points` includes only points inside the geometry.
+        boundary_points: A 2-D NumPy array.
+        boundary_normals: A 2-D NumPy array.
     """
 
-    def __init__(self, points):
+    def __init__(self, points, boundary_points=None, boundary_normals=None):
         self.points = np.asarray(points, dtype=config.real(np))
-        self.num_points = len(points)
+        self.num_points = len(points)
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:21]: Format via black.

File: deepxde/geometry/pointcloud.py [hunk 26]
```diff
@@ -9,25 +9,51 @@ class PointCloud(Geometry):
...
+                        "the shape of boundary_normals should be the same as boundary_points"
+                    )
+                self.boundary_normals = np.asarray(boundary_normals, dtype=config.real(np))
         super().__init__(
-            len(points[0]),
-            (np.amin(self.points, axis=0), np.amax(self.points, axis=0)),
-            np.inf,
-        )
-
+                    len(points[0]),
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:39]: Indent here is also wrong.

File: deepxde/geometry/pointcloud.py [hunk 27]
```diff
@@ -9,25 +9,67 @@ class PointCloud(Geometry):
...
+                    )
+                self.boundary_normals = np.asarray(
+                    boundary_normals, dtype=config.real(np)
+                )
         super().__init__(
             len(points[0]),
-            (np.amin(self.points, axis=0), np.amax(self.points, axis=0)),
+            (
+                np.amin(all_points, axis=0),
+                np.amax(all_points, axis=0),
```
Comment (lululxvi) [inline deepxde/geometry/pointcloud.py:46]: Delete the last "," and then format via black.

**Pull request merged**